sympy/simplify/combsimp.py

```diff
@@ -64,450 +47,70 @@ def combsimp(expr):
-        #
-        # run of 2 with args differing by 1/2
-        #
-        # >>> combsimp(gamma(x)*gamma(x+S.Half))
-        # 2*sqrt(2)*2**(-2*x - 1/2)*sqrt(pi)*gamma(2*x)
-        #
-        # run of 3 args differing by 1/3 (mod 1)
-        #
-        # >>> combsimp(gamma(x)*gamma(x+S(1)/3)*gamma(x+S(2)/3))
-        # 6*3**(-3*x - 1/2)*pi*gamma(3*x)
-        # >>> combsimp(gamma(x)*gamma(x+S(1)/3)*gamma(x+S(5)/3))
-        # 2*3**(-3*x - 1/2)*pi*(3*x + 2)*gamma(3*x)
-        #
-        def _run(coeffs):
-            # find runs in coeffs such that the difference in terms (mod 1)
-            # of t1, t2, ..., tn is 1/n
-            u = list(uniq(coeffs))
-            for i in range(len(u)):
-                dj = ([((u[j] - u[i]) % 1, j) for j in range(i + 1, len(u))])
-                for one, j in dj:
-                    if one.p == 1 and one.q != 1:
-                        n = one.q
-                        got = [i]
-                        get = list(range(1, n))
-                        for d, j in dj:
-                            m = n*d
-                            if m.is_Integer and m in get:
-                                get.remove(m)
-                                got.append(j)
-                                if not get:
-                                    break
-                        else:
-                            continue
-                        for i, j in enumerate(got):
-                            c = u[j]
-                            coeffs.remove(c)
-                            got[i] = c
-                        return one.q, got[0], got[1:]
-
-        def _mult_thm(gammas, numer, denom):
-            # pull off and analyze the leading coefficient from each gamma arg
-            # looking for runs in those Rationals
-
-            # expr -> coeff + resid -> rats[resid] = coeff
-            rats = {}
-            for g in gammas:
-                c, resid = g.as_coeff_Add()
-                rats.setdefault(resid, []).append(c)
-
-            # look for runs in Rationals for each resid
-            keys = sorted(rats, key=default_sort_key)
-            for resid in keys:
-                coeffs = list(sorted(rats[resid]))
-                new = []
-                while True:
-                    run = _run(coeffs)
-                    if run is None:
+                    nd_fact_args[1].extend([k.args[0]]*-rvd[k])
+                rvd[k] = 0
+        if not nd_fact_args[0] or not nd_fact_args[1]:
+            return rv
+
+        hit = False
+        for m in range(2):
+            i = 0
+            while i < len(nd_fact_args[m]):
+                ai = nd_fact_args[m][i]
+                for j in range(i + 1, len(nd_fact_args[m])):
+                    aj = nd_fact_args[m][j]
+
+                    sum = ai + aj
+                    if sum in nd_fact_args[1 - m]:
+                        hit = True
+
+                        nd_fact_args[1 - m].remove(sum)
+                        del nd_fact_args[m][j]
```

Comment:
Deleting items from a list while looping may lead to errors. Setting them to None was probably safer.

Reply:
@jksuom I thought it's safe since it is guaranteed that `j > i`, broken after `del`, and does not use list iterator. Or does it depends on the implementation of Python?

Reply:
I'm not sure. Maybe it will always work as `len(nd_fact_args[m])` is recomputed for each loop index `i`.